_posts/2022-10-05-#26114.md

```diff
@@ -0,0 +1,84 @@
+---
+layout: pr
+date: 2022-10-05
+title: "Make AddrFetch connections to fixed seeds"
+pr: 26114
+authors: [mzumsande]
+components: ["p2p"]
+host: glozow
+status: upcoming
+commit: 7e95f8f6bb
+
+---
+
+## Notes
+
+- Nodes connect to peers based on a set of addresses they know about, stored in their
+  [`AddrMan`](https://github.com/bitcoin/bitcoin/blob/f59e91511a3aa8b2770eeec7034ddc1a9dec918b/src/addrman.h#L86),
+but a brand new node starts out with an empty AddrMan. In order to participate in the network, the
+node needs to find addresses of peers to connect to.
+
+- Care is taken to not create a bias in the source of addresses. For example, it would be
+  inappropriate to have all new nodes connect to a list of 10 hard-coded addresses. That list of
+nodes would become outdated very quickly, have an enormous burden of serving blocks to new nodes,
+and be an eclipse vulnerability.
+
+	- If the node's addrman is empty, it queries
+[DNS seeds](https://github.com/bitcoin/bitcoin/blob/f59e91511a3aa8b2770eeec7034ddc1a9dec918b/src/chainparams.cpp#L118-L131)
+run by community members.
+
+	- If addrman is still empty, the node attempts connections to a set of fixed seeds,
+	  hard-coded and updated once per release. You can read more about how the fixed seeds are
+	  generated [here](https://github.com/bitcoin/bitcoin/tree/master/contrib/seeds).
+
+	- The node doesn't store addresses for
+	  [unreachable](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net.h#L211)
+	  networks. For example, if the node is only running on Tor, it won't add a clearnet address
+	  to its AddrMan.
+
+- Prior to [PR #26114](https://github.com/bitcoin/bitcoin/pull/26114), fixed seeds are added to
+  AddrMan. This means the node's first outbound connections are likely all taken from the fixed seeds.
+
+- An
+  [AddrFetch](https://github.com/bitcoin/bitcoin/blob/f59e91511a3aa8b2770eeec7034ddc1a9dec918b/src/node/connection_types.h#L70-L76)
+connection is a temporary outbound connection, used specifically for soliciting addresses.
+
+- The networking logic is multi-threaded (see list of threads and descriptions
+  [here](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads)).
+
+	- The `ThreadOpenConnections()` thread processes and sends messages to this node's peers.
+
+	- The `ThreadDNSAddressSeed` thread makes connections to DNS seeds.
+
+	- This PR also moves handling of fixed seeds from `ThreadOpenConnections` to
+	  `ThreadDNSAddressSeed`, and renames the latter to `ThreadAddressSeed`. Note this is
+	more involved than refactoring some code from one function to another; moving logic from one
+	thread to another could mean that data structures previously accessed by one thread are now
+	shared between multiple.
```

Comment:
Would it be useful to mention that we also need to be more careful about order of operations across threads and data races?